### ISSUE TITLE
Individuals without https://www.w3.org/2002/07/owl#NamedIndividual

### ISSUE THREAD

**Issue by Demirrr** [dice-group/Ontolearn#324]
Without `x rdf:type https://www.w3.org/2002/07/owl#NamedIndividual`, we are unable to consider  such `x` as a positive nor negative example.

If we know that `x` has at least a single type info ( e.g. `x rdf:type Person`), can we assume that `x rdf:type https://www.w3.org/2002/07/owl#NamedIndividual` must hold.

**Comment by Jean-KOUAGOU** [comment]
OwlNamedIndividual is not supposed to be a class, right?

**Comment by Jean-KOUAGOU** [comment]
Why do we need `x rdf:type https://www.w3.org/2002/07/owl#NamedIndividual`?

**Comment by Demirrr** [comment]
owl:NamedIndividual is a class.
```
owl:NamedIndividual a rdfs:Class ;
     rdfs:label "NamedIndividual" ;
     rdfs:comment "The class of named individuals." ;
     rdfs:isDefinedBy <http://www.w3.org/2002/07/owl#> ;
     rdfs:subClassOf owl:Thing .
```
see https://www.w3.org/2002/07/owl#NamedIndividual for more

Without this explicit info `x rdf:type https://www.w3.org/2002/07/owl#NamedIndividual` about `x`, Knowledgebase() class does consider `x` as an individual. 

Long story short:


`print(KnowledgeBase(path="Family"))`

Remove a triple involving `rdf:type https://www.w3.org/2002/07/owl#NamedIndividual` from Family dataset and observe that number of individuals is decreased by one and this individual cannot be used as a positive or negative example.

`print(KnowledgeBase(path="Family"))`

**Comment by alkidbaci** [comment]
Its true that if the triple `rdf:type https://www.w3.org/2002/07/owl#NamedIndividual` does not exist for a individual it will not be recognized as such by individual retrieval methods which of course can be a problem but since concept learners mainly use reasoner for their processes, the `instance` method will recognize these as individuals of a certain type, e.g. `Person` and the learning problem examples are still being considered at some extent. As much I know there can be classes that are depended on the individuals retrieval methods and `instances` method so in that case this is still a problem.

**Comment by Demirrr** [comment]
> Its true that if the triple rdf:type https://www.w3.org/2002/07/owl#NamedIndividual does not exist for a individual it will not be recognized as such by individual retrieval methods which of course can be a problem but since concept learners mainly use reasoner for their processes

:thumbsup:


>  the instance method will recognize these as individuals of a certain type, e.g. Person and the learning problem examples are still being considered at some extent

Not quite sure whether I understood this statement. "these" refer to DL individuals without explicit type information (i.e., `x rdf:type https://www.w3.org/2002/07/owl#NamedIndividual`) right ? 

I reckon that considering `x` is a DL individual or not should be done by a reasoner. My intiutive respond is to consider 
x as a DL individual if we know that x has at least a single type info.

**Comment by Demirrr** [comment]
Oh I see. Thank you for the explanation.

Do you think whether we could/should introduce a boolean attribute for Knowledgebase class that allows users to choose `individuals_in_signature`  or `individuals`, hence the size of the `individuals` must be always greater or requal to the size of  `individuals_in_signature`?

Is there any disadtange of introducing this boolean attribute ?

**Comment by alkidbaci** [comment]
I don't think that will make any difference. `individuals` is a method of KnowledeBase which is related with `individuals_in_signature ` of its Ontology object. When being asked for all individuals everything leads to `individuals_in_signature`, but you can ask for individuals of a specific class expression and than it will use `instances` method of the reasoner. Now the idea is basically the same, behind the latter 2 methods - you can get all the individuals if you ask for `instances` of the class `OWLNamedIndividuals`. 

If the goal is to make identification of all individuals regardless of them contain the triple `x rdf:type https://www.w3.org/2002/07/owl#NamedIndividual` this is kind of chicken or egg problem. How can we distinct an individual without knowing it is of type `https://www.w3.org/2002/07/owl#NamedIndividual`? Isn't it a strict rule for individuals to explicitly be of that type?

**Comment by Demirrr** [comment]
Here is a solution 

### ontolearn 0.6.1. 

```python
kb=Knowledgebase(path="..")
print(kb) # KnowledgeBase(path='KGs/Family/family-benchmark_rich_background.owl' <18 classes, 4 properties, 0 individuals)

for i in kb.get_concepts():
     print(len(list(kb.individuals(i))) # 10, 20, 0 etc.
```
Here, we provide inconsistent/incorrect knowledge to users by saying that **there is no individual found but if you ask individuals of concepts you get one**.

### ontolearn 0.6.2. 

```python
kb=Knowledgebase(path="..")
print(kb) # KnowledgeBase(path='KGs/Family/family-benchmark_rich_background.owl' <18 classes, 4 properties, x individuals)
# where x is computed as 

from itertools import chain
individuals=set(chain.from_iterable(kb.individuals(i) for i in kb.get_concepts()))
x = len(individuals)
```


Can you implement this solution ?

**Comment by alkidbaci** [comment]
I recorded the time it takes to load carcinogenesis dataset which has 22372 individuals in the KnowledgeBase:
- with FIC reasoner which is the default reasoner (access inds directly from FIC attributes): **`2.04`**  sec
- with another reasoner: **`2.21 ~ 2.27`** sec (uses the method `individuals_in_signature`)
- with solution of ontolearn 0.6.2: **`2.68 ~ 2.77`** sec 

> we could/should introduce a boolean attribute for Knowledgebase class

Considering the results and the idea that you gave earlier, should I make the ontolearn 0.6.2 solution optional, depending on a bool argument?

**Comment by Demirrr** [comment]
Yes please
